pkg/kvstore/etcd.go

```diff
@@ -708,6 +703,20 @@ func makeSessionName(sessionPrefix string, opts *ExtraOptions) string {
 	return sessionPrefix
 }
 
+func (e *etcdClient) waitForReadyConnection(ctx context.Context) error {
+	for {
+		state := e.client.ActiveConnection().GetState()
```

Comment:
hmm, it's very non-obvious to me why the connection wouldn't be in an IDLE state. It means that something somewhere had to issue an RPC call as per GRPC documentation:
```
IDLE: This is the state where the channel is not even trying to create a connection because of a lack of new or pending RPCs. New RPCs MAY be created in this state. Any attempt to start an RPC on the channel will push the channel out of this state to connecting. When there has been no RPC activity on a channel for a specified IDLE_TIMEOUT, i.e., no new or pending (active) RPCs for this period, channels that are READY or CONNECTING switch to IDLE. Additionally, channels that receive a GOAWAY when there are no active or pending RPCs should also switch to IDLE to avoid connection overload at servers that are attempting to shed connections. We will use a default IDLE_TIMEOUT of 300 seconds (5 minutes).
```
[source](https://grpc.github.io/grpc/core/md_doc_connectivity-semantics-and-api.html)
So essentially waiting here on Ready state AFAIK depends on this call in newClient AFAIK:
https://github.com/cilium/cilium/blob/09872a6758f52b40f9d540e076bbf1eef1100579/vendor/go.etcd.io/etcd/client/v3/client.go#L450

but then the question is, is it worth doing that if newClient already performs that call and would return an error if RPC failed?

Reply:
My understanding is that the connection transitions out of idle mode due to the `newClient` calling `dialWithBalancer`, which eventually calls `grpc.DialContext`:

https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/vendor/go.etcd.io/etcd/client/v3/client.go#L429

https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/vendor/go.etcd.io/etcd/client/v3/client.go#L318

According to the `DialContext`'s godoc:

https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/vendor/google.golang.org/grpc/clientconn.go#L209-L211

Given that the goal here is to wait for the connection to be established (with a timeout), one possibility could be to set the `WithBlock` dial option, and configure a non-zero DialTimeout. However, `WithBlock` is [deprecated and considered an anti-pattern](https://github.com/grpc/grpc-go/blob/master/Documentation/anti-patterns.md). The documentation suggests to use `WaitForStateChange` instead.

Differently, `getToken` should not have effects, because neither the username nor the password are set as we use certs for authentication:

https://github.com/cilium/cilium/blob/7e9de944a06eca20a88abd8d072d751dd3b56447/vendor/go.etcd.io/etcd/client/v3/client.go#L272-L277

I guess that yet another possibility (maybe the best one actually) could be removing the `waitForReadyConnection` step and just wait for the `ListDone` event on the heartbeat prefix. That would ensure that we are actually connected, and not depend on the underlying grpc connection status (which may indeed turn idle again after a timeout if no requests are issued). WDYT?

Reply:
Yep, I agree. Let me change that and verify that it works as expected.

Reply:
I did a bit of testing, and it seemed to work great :rocket: